klippy/extras/palette2.py

```diff
@@ -0,0 +1,610 @@
+                "Unable to auto load filament, extruder is below minimum temp")
+            return
+
+        if self.smart_load_timer is None:
+            logging.info("Smart load starting")
+            reactor = self.printer.get_reactor()
+            self.smart_load_timer = reactor.register_timer(
+                self._run_Smart_Load, reactor.NOW)
+
+    def cmd_P2(self, line):
+        t = line.split()
+        ocode = t[0]
+        params = t[1:]
+        params_count = len(params)
+        if params_count:
+            res = [i for i in params if i[0] == "D" or i[0] == "U"]
+            if not all(res):
+                logging.error("Omega parameters are invalid")
+                return
+
+        func = getattr(self, 'cmd_P2_' + ocode, None)
+        if func is not None:
+            func(params)
+
+    def _param_Matcher(self, matchers, params):
+        # Match the command with the handling table
+        for matcher in matchers:
+            if len(params) >= matcher[1]:
+                match_params = matcher[2:]
+                res = all([match_params[i] == params[i]
+                           for i in range(len(match_params))])
+                if res:
+                    matcher[0](params)
+                    return True
+
+        return False
+
+    def _run_Read(self, eventtime):
+        while self.serial.in_waiting > 0:
+            try:
+                raw_line = self.serial.readline()
+            except SerialException:
+                logging.error("Unable to communicate with the Palette 2")
+                self.cmd_Disconnect()
+                return eventtime + SERIAL_TIMER
+
+            if not len(raw_line):
+                break
+
+            text_line = raw_line.decode().strip()
+            if not len(text_line):
+                break
+
+            # Line was return without timeout
+            logging.debug("%s P2 -> : %s" % (time.time(), text_line))
+
+            # Received a heartbeat from the device
+            if text_line == COMMAND_HEARTBEAT:
+                self.heartbeat = time.time()
+
+            elif text_line[0] == "O":
+                self.cmd_P2(text_line)
+
+        # do a heartbeat check
+        if self.heartbeat and self.heartbeat < (
+                time.time() - HEARTBEAT_TIMEOUT):
+            logging.error(
+                "P2 has not responded to heartbeat, Palette will disconnect")
+            self.cmd_Disconnect()
+        return eventtime + SERIAL_TIMER
+
+    def _run_Write(self, eventtime):
+        while True:
+            try:
+                text_line = self.write_queue.get_nowait()
+            except Empty:
+                break
+
+            if text_line:
+                self.omega_last_command = text_line
+                l = text_line.strip()
+                logging.debug("%s -> P2 : %s" % (time.time(), l))
+                terminated_line = "%s\n" % (l)
+
+                try:
+                    self.serial.write(terminated_line.encode())
```

Comment:
Same issue here - this call could block indefinitely.

Reply:
I will need to check, but the timeout should apply here too. Gauranteed to return in 500ms. Same comment as above though, would you prefer threads?

Reply:
Moves to single threaded fully non-blocking. Serial sets **timeout=0**, **write_timeout=0** in **cmd_Connect**